File: scripts/text_to_video.py
#!/usr/bin/env python3
"""
Text-to-video pipeline: FLUX (text→image) + SVD (image→video)

Requires both services to be available (run sequentially, not simultaneously).
"""

import argparse
import requests
import time
import random
import sys
from pathlib import Path

FLUX_SERVER = "http://192.168.1.64:8570"
SVD_SERVER = "http://192.168.1.64:8100"


def check_server(url: str, name: str) -> bool:
    """Check if a server is responding."""
    try:
        r = requests.get(f"{url}/system_stats", timeout=5)
        return r.status_code == 200
    except:
        return False


def generate_image(prompt: str, width: int, height: int, seed: int) -> bytes:
    """Generate image with FLUX."""
    workflow = {
        "prompt": {
            "1": {"class_type": "CheckpointLoaderSimple",
                  "inputs": {"ckpt_name": "flux1-schnell-fp8.safetensors"}},
            "2": {"class_type": "CLIPTextEncode",
                  "inputs": {"text": prompt, "clip": ["1", 1]}},
            "3": {"class_type": "CLIPTextEncode",
                  "inputs": {"text": "", "clip": ["1", 1]}},
            "4": {"class_type": "EmptyLatentImage",
                  "inputs": {"width": width, "height": height, "batch_size": 1}},
            "5": {"class_type": "KSampler",
                  "inputs": {"model": ["1", 0], "positive": ["2", 0],
                             "negative": ["3", 0], "latent_image": ["4", 0],
                             "seed": seed, "steps": 4, "cfg": 1.0,
                             "sampler_name": "euler", "scheduler": "simple",
                             "denoise": 1.0}},
            "6": {"class_type": "VAEDecode",
                  "inputs": {"samples": ["5", 0], "vae": ["1", 2]}},
            "7": {"class_type": "SaveImage",
                  "inputs": {"images": ["6", 0], "filename_prefix": "t2v"}},
        }
    }

    r = requests.post(f"{FLUX_SERVER}/prompt", json=workflow)
    r.raise_for_status()
    prompt_id = r.json()["prompt_id"]

    # Poll
    while True:
        r = requests.get(f"{FLUX_SERVER}/history/{prompt_id}")
        data = r.json()
        if prompt_id in data and data[prompt_id].get("status", {}).get("completed"):
            break
        print(".", end="", flush=True)
        time.sleep(2)

    # Get filename and download
    outputs = data[prompt_id]["outputs"]
    for node_out in outputs.values():
        if "images" in node_out:
            filename = node_out["images"][0]["filename"]
            break

    r = requests.get(f"{FLUX_SERVER}/view", params={
        "filename": filename, "subfolder": "", "type": "output"
    })
    r.raise_for_status()
    return r.content


def generate_video(image_bytes: bytes, width: int, height: int,
                   frames: int, motion: int, fps: int, seed: int) -> bytes:
    """Generate video with SVD from image bytes."""

    # Upload image
    r = requests.post(
        f"{SVD_SERVER}/upload/image",
        files={"image": ("input.png", image_bytes, "image/png")},
        data={"subfolder": "", "type": "input"},
    )
    r.raise_for_status()
    image_name = r.json()["name"]

    workflow = {
        "prompt": {
            "1": {"class_type": "ImageOnlyCheckpointLoader",
                  "inputs": {"ckpt_name": "svd_xt.safetensors"}},
            "2": {"class_type": "LoadImage",
                  "inputs": {"image": image_name}},
            "3": {"class_type": "SVD_img2vid_Conditioning",
                  "inputs": {
                      "clip_vision": ["1", 1], "init_image": ["2", 0],
                      "vae": ["1", 2], "width": width, "height": height,
                      "video_frames": frames,
                      "motion_bucket_id": motion,
                      "fps": fps, "augmentation_level": 0.0,
                  }},
            "4": {"class_type": "KSampler",
                  "inputs": {
                      "model": ["1", 0], "positive": ["3", 0],
                      "negative": ["3", 1], "latent_image": ["3", 2],
                      "seed": seed, "steps": 20, "cfg": 2.5,
                      "sampler_name": "euler", "scheduler": "karras",
                      "denoise": 1.0,
                  }},
            "5": {"class_type": "VAEDecode",
                  "inputs": {"samples": ["4", 0], "vae": ["1", 2]}},
            "6": {"class_type": "VHS_VideoCombine",
                  "inputs": {
                      "images": ["5", 0], "frame_rate": fps,
                      "loop_count": 0, "filename_prefix": "t2v",
                      "format": "video/h264-mp4", "pingpong": False,
                      "save_output": True, "pix_fmt": "yuv420p",
                      "crf": 19, "save_metadata": True,
                      "trim_to_audio": False,
                  }},
        }
    }

    r = requests.post(f"{SVD_SERVER}/prompt", json=workflow)
    r.raise_for_status()
    prompt_id = r.json()["prompt_id"]

    # Poll
    while True:
        r = requests.get(f"{SVD_SERVER}/history/{prompt_id}")
        data = r.json()
        if prompt_id in data and data[prompt_id].get("status", {}).get("completed"):
            break
        print(".", end="", flush=True)
        time.sleep(5)

    # Get filename and download
    outputs = data[prompt_id]["outputs"]
    for node_out in outputs.values():
        if "gifs" in node_out:
            filename = node_out["gifs"][0]["filename"]
            break

    r = requests.get(f"{SVD_SERVER}/view", params={
        "filename": filename, "subfolder": "", "type": "output"
    })
    r.raise_for_status()
    return r.content


def main():
    parser = argparse.ArgumentParser(description="Text to video via FLUX + SVD")
    parser.add_argument("--prompt", "-p", required=True, help="Text prompt")
    parser.add_argument("--output", "-o", required=True, help="Output video path")
    parser.add_argument("--orientation", "-r", choices=["portrait", "landscape", "square"],
                       default="portrait", help="Video orientation")
    parser.add_argument("--duration", "-d", type=float, default=5.0, help="Duration in seconds")
    parser.add_argument("--fps", type=int, default=6, help="Frame rate")
    parser.add_argument("--motion", "-m", type=int, default=80, help="Motion amount 1-300")
    parser.add_argument("--seed", "-s", type=int, help="Random seed")

    args = parser.parse_args()
    seed = args.seed or random.randint(0, 2**32)
    frames = int(args.duration * args.fps)

    # Resolution based on orientation
    if args.orientation == "portrait":
        width, height = 576, 1024  # 9:16
    elif args.orientation == "landscape":
        width, height = 1024, 576  # 16:9
    else:
        width, height = 768, 768

    print(f"=== Text-to-Video Pipeline ===")
    print(f"Prompt: {args.prompt[:60]}...")
    print(f"Size: {width}x{height} ({args.orientation})")
    print(f"Duration: {args.duration}s ({frames} frames @ {args.fps}fps)")
    print(f"Motion: {args.motion}")
    print(f"Seed: {seed}")
    print()

    # Check which server is up
    flux_up = check_server(FLUX_SERVER, "FLUX")
    svd_up = check_server(SVD_SERVER, "SVD")

    print(f"FLUX (text-to-image): {'✓ online' if flux_up else '✗ offline'}")
    print(f"SVD (image-to-video): {'✓ online' if svd_up else '✗ offline'}")
    print()

    if not flux_up and not svd_up:
        print("ERROR: Neither server is running!")
        sys.exit(1)

    if flux_up and svd_up:
        print("Both servers online - running full pipeline...")
    elif flux_up:
        print("Only FLUX online - will generate image only")
    else:
        print("Only SVD online - need FLUX for text-to-image step")
        print("Start FLUX server and try again, or provide an image with svd_client.py")
        sys.exit(1)

    # Step 1: Generate image
    print("\n[1/2] Generating image with FLUX", end="", flush=True)
    start = time.time()
    image_bytes = generate_image(args.prompt, width, height, seed)
    print(f" done ({time.time()-start:.1f}s)")

    # Save intermediate image
    img_path = Path(args.output).with_suffix(".png")
    with open(img_path, "wb") as f:
        f.write(image_bytes)
    print(f"  Saved: {img_path}")

    if not svd_up:
        print("\nSVD offline - image saved, video skipped")
        return

    # Step 2: Generate video
    print("\n[2/2] Generating video with SVD", end="", flush=True)
    start = time.time()
    video_bytes = generate_video(image_bytes, width, height, frames, args.motion, args.fps, seed)
    print(f" done ({time.time()-start:.1f}s)")

    with open(args.output, "wb") as f:
        f.write(video_bytes)
    print(f"  Saved: {args.output}")


if __name__ == "__main__":
    main()

File: components/vwf-engine/crates/vwf-steps/src/text_to_video.rs
//! Handler for text_to_video step kind.
//!
//! Generates video from a text prompt using Wan 2.2 via ComfyUI REST API.

use anyhow::{Context, Result};
use serde::Deserialize;
use serde_json::Value;
use std::process::Command;

use super::context::StepCtx;

#[derive(Deserialize)]
struct Payload {
    /// Text prompt for video generation
    prompt: String,
    /// Output video path (.mp4)
    output_path: String,
    /// Video orientation: portrait, landscape, square (default: landscape)
    #[serde(default = "default_orientation")]
    orientation: String,
    /// Number of frames to generate (default: 81 = ~5s at 16fps)
    #[serde(default = "default_length")]
    length: u32,
    /// Sampling steps (default: 20)
    #[serde(default = "default_steps")]
    steps: u32,
    /// CFG scale (default: 5.0)
    #[serde(default = "default_cfg")]
    cfg: f32,
    /// Random seed for reproducibility
    seed: Option<u64>,
    /// ComfyUI server URL
    #[serde(default = "default_server")]
    server: String,
    /// Python interpreter path (default: python3)
    python_path: Option<String>,
}

fn default_orientation() -> String {
    "landscape".to_string()
}
fn default_length() -> u32 {
    81
}
fn default_steps() -> u32 {
    20
}
fn default_cfg() -> f32 {
    5.0
}
fn default_server() -> String {
    "http://192.168.1.64:6000".to_string()
}

pub fn execute(ctx: &mut StepCtx<'_>, payload: &Value) -> Result<()> {
    let p: Payload = serde_json::from_value(payload.clone())
        .with_context(|| ctx.error_context("payload decode text_to_video"))?;

    let prompt = ctx.render(&p.prompt)?;
    let output_path = ctx.render(&p.output_path)?;
    let server = ctx.render(&p.server)?;
    let orientation = ctx.render(&p.orientation)?;

    // Resolution presets (latent dimensions - output is 2x)
    let (width, height) = match orientation.as_str() {
        "portrait" => (480, 832), // Output: 960x1664
        "square" => (640, 640),   // Output: 1280x1280
        _ => (832, 480),          // Output: 1664x960 (landscape)
    };

    let seed = p.seed.unwrap_or_else(rand::random);
    let python = p
        .python_path
        .as_ref()
        .map(|pp| ctx.render(pp))
        .transpose()?
        .unwrap_or_else(|| "python3".to_string());

    let script = video_gen_script(
        &server,
        &prompt,
        &output_path,
        width,
        height,
        p.length,
        p.steps,
        p.cfg,
        seed,
    );

    let status = Command::new(&python)
        .args(["-c", &script])
        .status()
        .with_context(|| ctx.error_context("spawn text_to_video python"))?;

    if !status.success() {
        anyhow::bail!(
            "Video generation failed with exit code: {:?}",
            status.code()
        );
    }

    Ok(())
}

#[allow(clippy::too_many_arguments)]
fn video_gen_script(
    server: &str,
    prompt: &str,
    output: &str,
    width: u32,
    height: u32,
    length: u32,
    steps: u32,
    cfg: f32,
    seed: u64,
) -> String {
    let prompt_escaped = prompt.replace('\\', "\\\\").replace('"', "\\\"");
    let negative = "blurry, low quality, distorted, watermark, text, deformed";

    format!(
        r#"
import requests
import time
import sys
from pathlib import Path

SERVER = "{server}"
PROMPT = "{prompt_escaped}"
NEGATIVE = "{negative}"
OUTPUT = "{output}"
WIDTH = {width}
HEIGHT = {height}
LENGTH = {length}
STEPS = {steps}
CFG = {cfg}
SEED = {seed}

workflow = {{
    "prompt": {{
        "1": {{"class_type": "UNETLoader",
              "inputs": {{"unet_name": "wan2.2_ti2v_5B_fp16.safetensors",
                         "weight_dtype": "fp8_e4m3fn"}}}},
        "2": {{"class_type": "CLIPLoader",
              "inputs": {{"clip_name": "umt5_xxl_fp8_e4m3fn_scaled.safetensors",
                         "type": "wan"}}}},
        "3": {{"class_type": "VAELoader",
              "inputs": {{"vae_name": "wan2.2_vae.safetensors"}}}},
        "4": {{"class_type": "CLIPTextEncode",
              "inputs": {{"text": PROMPT, "clip": ["2", 0]}}}},
        "5": {{"class_type": "CLIPTextEncode",
              "inputs": {{"text": NEGATIVE, "clip": ["2", 0]}}}},
        "6": {{"class_type": "WanImageToVideo",
              "inputs": {{
                  "positive": ["4", 0], "negative": ["5", 0],
                  "vae": ["3", 0], "width": WIDTH, "height": HEIGHT,
                  "length": LENGTH, "batch_size": 1}}}},
        "7": {{"class_type": "KSampler",
              "inputs": {{
                  "model": ["1", 0], "positive": ["6", 0],
                  "negative": ["6", 1], "latent_image": ["6", 2],
                  "seed": SEED, "steps": STEPS, "cfg": CFG,
                  "sampler_name": "euler", "scheduler": "normal",
                  "denoise": 1.0}}}},
        "8": {{"class_type": "VAEDecode",
              "inputs": {{"samples": ["7", 0], "vae": ["3", 0]}}}},
        "9": {{"class_type": "VHS_VideoCombine",
              "inputs": {{
                  "images": ["8", 0], "frame_rate": 16,
                  "loop_count": 0, "filename_prefix": "wan22",
                  "format": "video/h264-mp4", "pingpong": False,
                  "save_output": True}}}},
    }}
}}

print(f"Submitting to {{SERVER}}...")
print(f"  Prompt: {{PROMPT[:60]}}{{'...' if len(PROMPT) > 60 else ''}}")
print(f"  Latent: {{WIDTH}}x{{HEIGHT}} -> Output: {{WIDTH*2}}x{{HEIGHT*2}}")
print(f"  Frames: {{LENGTH}} ({{LENGTH/16:.1f}}s @ 16fps)")

r = requests.post(f"{{SERVER}}/prompt", json=workflow)
r.raise_for_status()
response = r.json()

if response.get("node_errors"):
    print(f"ERROR: {{response['node_errors']}}")
    sys.exit(1)

prompt_id = response["prompt_id"]
print(f"  Job ID: {{prompt_id}}")

print("  Generating (this takes ~13 minutes for 81 frames)", end="", flush=True)
while True:
    r = requests.get(f"{{SERVER}}/history/{{prompt_id}}")
    data = r.json()
    if prompt_id in data and data[prompt_id].get("status", {{}}).get("completed"):
        break
    print(".", end="", flush=True)
    time.sleep(30)
print(" done")

# Extract filename
outputs = data[prompt_id]["outputs"]
filename = None
for node_out in outputs.values():
    if "gifs" in node_out:
        filename = node_out["gifs"][0]["filename"]
        break

if filename is None:
    raise RuntimeError(f"No video output found in: {{outputs}}")

print(f"  Downloading: {{filename}}")
r = requests.get(f"{{SERVER}}/view", params={{"filename": filename, "subfolder": "", "type": "output"}})
r.raise_for_status()

Path(OUTPUT).parent.mkdir(parents=True, exist_ok=True)
with open(OUTPUT, "wb") as f:
    f.write(r.content)
print(f"Saved: {{OUTPUT}}")
"#
    )
}
